09-intermediate-rails/session.md

```diff
@@ -0,0 +1,117 @@
+
+Now let's see what we end up with!
+
+## `flash.now`
+
+The above example works because right after we set the `flash` value in the controller, we are doing a `redirect_to`. If we want to use a `flash` but with a `render` instead of a `redirect_to` then we need to use `flash.now` instead. The different usage is solely based on what comes next. `flash.now` will only stick around going to a `render` and `flash` will stick around going to a `redirect`.
+
+## `session`
+`session` is another special type of hash-like-object that we utilize to keep track of data throughout a users "session", which normally ends when they close their browser. It is similar to `params` in that it is something special that Rails creates for us. It is different because it persists from one one request to another. It is most often used to store information about a user when they log in.
+
+Rails will automatically delete the information from the `session` when the user closes the browser. We can also manually remove the data if we not longer want to track it (like when the user logs out).
+
+Let's see it! We are going to create a special form which is going to allow our authors to "log in". This is not a secure way to simulate logging in, but we are using this as our first stepping stone to creating "real" authentication, which we'll do later on. We'll allow the user to type an authors name which will "log in" as that author.
+
+We'll need to a create a few new things in our application in order to get set up for this:
+1. Create a login form and form submission route (think of how we do `new` and `create`)
+2. Create a special controller for managing this login process called `SessionsController`
+3. Add empty controller actions for the routes in step #1
+4. Create a form view where our user will enter in the author's name
+5. Create our initial controller action which will set the `session`
+```ruby
+# app/controllers/sessions_controller.rb
+...
+  def login
+    author = Author.find_by(name: params[:name])
+
+    if author
+      session[:author_id] = author.id
+    end
+  end
+```
+
+6. Update the controller action to improve the user experience after they log in:
+```ruby
+# app/controllers/sessions_controller.rb
+...
+  def login
+    author = Author.find_by(name: params[:name])
+
+    if author
+      session[:author_id] = author.id
+      flash[:success] = "#{ author.name } is successfully logged in"
+      redirect_to root_path
+    end
+  end
+```
+
+**Question**: Now where can we utilize this logged in Author information?  
+**Question**: How you think you'd use the `session` in a logout action?
+
+## Key Takeaways
+Rails provides a few special hash-like objects that allow us to go above and beyond local and instance variables in our Rails applications.
+
+We will utilize `flash` and `flash.now` to provide additional feedback to our users as they use our applications.
+
+We will utilize `session` to keep track of logged in user information.
```

Comment:
Is it worthwhile making the comparison to `params`? It's another hash-like object provided by Rails, but it operates in a totally different way.